modules/core/src/test/java/org/apache/ignite/internal/processors/cache/GridCacheRebalanceOrderTest.java

```diff
@@ -0,0 +1,177 @@
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.internal.processors.cache;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import org.apache.ignite.Ignite;
+import org.apache.ignite.IgniteCache;
+import org.apache.ignite.cache.CacheRebalanceMode;
+import org.apache.ignite.configuration.CacheConfiguration;
+import org.apache.ignite.configuration.IgniteConfiguration;
+import org.apache.ignite.events.CacheRebalancingEvent;
+import org.apache.ignite.events.Event;
+import org.apache.ignite.events.EventType;
+import org.apache.ignite.internal.util.typedef.T2;
+import org.apache.ignite.lang.IgniteBiTuple;
+import org.apache.ignite.lang.IgnitePredicate;
+import org.apache.ignite.testframework.junits.common.GridCommonAbstractTest;
+import org.junit.After;
+import org.junit.Test;
+
+import static org.apache.ignite.cache.CacheRebalanceMode.ASYNC;
+import static org.apache.ignite.cache.CacheRebalanceMode.SYNC;
+import static org.apache.ignite.events.EventType.EVT_CACHE_REBALANCE_STOPPED;
+
+/**
+ * Tests that caches with rebalance mode equals to SYNC are rebalanced in the first place.
+ */
+public class GridCacheRebalanceOrderTest extends GridCommonAbstractTest {
+    /** Rebalance timeout. */
+    private static final long REBALANCE_TIMEOUT = 5_000;
+
+    /** Flag indicates that local listener should be used to track rebalance events. */
+    private boolean trackRebalanceEvts;
+
+    /** Caches rebalance events. */
+    private final List<CacheRebalancingEvent> evtsList = Collections.synchronizedList(new ArrayList<>());
+
+    /** {@inheritDoc} */
+    @Override protected IgniteConfiguration getConfiguration(String igniteInstanceName) throws Exception {
+        IgniteConfiguration cfg = super.getConfiguration(igniteInstanceName);
+
+        if (trackRebalanceEvts) {
+            Map<IgnitePredicate<? extends Event>, int[]> listeners = new HashMap<>();
+
+            listeners.put(new IgnitePredicate<CacheRebalancingEvent>() {
+                @Override public boolean apply(CacheRebalancingEvent evt) {
+                    evtsList.add(evt);
+
+                    return true;
+                }
+            }, new int[] {EVT_CACHE_REBALANCE_STOPPED});
```

Comment:
Do we have a strong guarantee that such events will be received by listeners with the same order they fired?

Reply:
If I am not mistaken, the answer is yes, we have such a guarantee. Unfortunately, it is not clearly stated in the Javadoc.